catalpainternational/canoe
# Minimal mode app booting

Closes #269 

### Related
Comes on top of #268 and #271 . Those can be examined and merged first, or disregarded and treated as part of this PR. But it makes more sense to examine & merge those separately & first,  because they address separate issues (bugs) that have no direct relation to the #269 Lean booting goal (but I needed to fix them, because crash.). They're bugfixes, good to have regardless of this here PR.

***Note that there are no user-visible changes for existing use cases***. I'm trying to do the minimal, and the minimal can be improved upon. Especially keep in mind that yes, while `U+1f638` is debatable, no user and probably no dev will see it in the user interface, at least not until they mess about a bunch with their devtools. Hum-ing and hehm-ing about `U+1f638` is not what I want this PR and its reviews to be about.

### What does it do

It makes it so that Canoe can boot with just a `Serviceworker` and some `Cache`s installed. This makes it easier for an eventual Appelflap state transfer implementation to take care of, well, transferring state, as it doesn't have to set project-specific cookies/localStorage/IDB stuff in project-specific ways just to poke things into a state that the webapp doesn't crash with or does something weird with. It should be the webapp's responsibility to be able to bootstrap itself from a minimal state, reinitializing things where necessary. And now Canoe can! Hurray again!

In the process, out of necessity, I also made the "are we authenticated" evidence-based rather than belief-based.
It used to be based solely on presence or absence of a cookie with name "Token", disregarding the payload.
And even if it would be verifying the payload (which it can't, because currently Canoe's tokens are HMAC based rather than signature based!), it wouldn't know whether the server hasn't rotated keys or disabled the user. Evidence-based is just better.

In a nutshell:
- Don't assume presence of a cookie with name "Token" means we're authenticated (and will always be authenticated)
- Don't set the user stuck on a login screen when no such "Token" cookie is available — the user might be offline, unable to login, but might want to interact with the learning content that's available from the cache.
- Track authentication status. We're authenticated until proven otherwise, and once proven deauthenticated, we're deauthenticated until we've logged in again.

### Testing

1. Load the webapp on a clean profile as usual
2. Log in to fill your caches with e-learning goodness
3. Set yourself to offline (via network tab of devtools)
4. Play around with deleting any/some/all of Cookies, IDB, LocalStorage items, ... But not Caches, or the Serviceworker, duh.
5. For any of the operations of step 4, witness that the webapp still boots (ctrl-L, enter)
6. Hop online, load the webapp (ctrl-L, enter)
7. If you had deleted the cookie, you should now see a login screen - that you didn't see when you were offline, which is good, because it was no good presenting a login screen to a user who's offline!


### Possible further improvements

#### Deauthentication detection

Currently the deauthentication-detection piggybacks on the requests to various backend-pages ("actions", lesson pages. I think). 

Nothing wrong with that, but I'm quite unfamiliar with this project, there may be better candidates that are more regularly polled. I don't know, is there anything that is ever polled again after boot? What if there are content updates? Is the user expected to refresh the page regularly or something?

In Gather I poll an endpoint on an interval, and this *invites* the user to re-login if it turns out that the user isn't anymore —
rather than dumping them to a login screen that can't be bypassed. Especially with intermittent connectivity, an invitation is the better approach — while there is a network available at the time of unauthenticatedness detection, there might not be one by the time the user walks back to their browser/tablet! So the poll-and-invite approach is clearly superior in that respect, especially if the invitation to re-login is automatically retracted when the network goes away (it's a stretch, but it'd be really good UX). In Gather, the re-login invitation is canceled if by some magic the authentication succeeds again, which is also nice. But in this app, with this PR, once a deauthentication is detected, you'll be dumped to the login screen (on reload anyway) with no way to restore access to the learning materials (that may actually be right there, on the device! how frustrating! locked out by your own devices! At least good oldfashioned books don't pull that trick on you!) other than having a) internet b) a valid login . For remote no-internet scenarios that's clearly undesirable. To fix that we'll need a wider discussion. I'd suggest a local-only-mode one-button on the login screen that just lets you get on with what you have on the device.


#### Auto-guest-login

Currently, as I'm just trying to do minimally invasive stuff to get to the desired goal of lean boots, no super smart behaviour is set up. But I saw that there was something like a "guest mode" on haroman.tl. On a cold boot of the webapp, the user could be offered to use the guest account instead of the rather minimal anonymous mode that's currently set up with this PR.
